### PR TITLE
[Feature] Support paimon materialized view

### LINKED ISSUE #30232: Support materialized view for paimon
## Feature request

Support materialized view for paimon table.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

Additional context

- [ ] Support unparttitioned paimon table

- [ ] Support partitioned paimon table

- [ ] Support basic query rewrite

- [ ] Support modfied partition refresh

- [ ] Support partial partition query rewrite

- [ ] Support materialized view delta join query rewrite

### PR DESCRIPTION
**Pull request by wg1026688210**
## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

fix #30232
Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 39 / 56 (69.64%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/PaimonTable.java|0|2|00.00%|[95, 100]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|4|8|50.00%|[266, 267, 268, 269]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|12|18|66.67%|[1225, 1231, 1235, 1241, 1242, 1243]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|16|21|76.19%|[180, 811, 867, 870, 873]|
|:large_blue_circle:|com/starrocks/connector/paimon/RowDataConverter.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 25 / 82 (30.49%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/PaimonPartitionKey.java|0|3|00.00%|[23, 24, 28]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|2|22|09.09%|[715, 821, 822, 825, 826, 827, 828, 829, 830, 832, 833, 834, 838, 883, 884, 885, 886, 887, 888, 889]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|4|21|19.05%|[1229, 1230, 1231, 1232, 1235, 1238, 1239, 1242, 1243, 1244, 1246, 1247, 1251, 1252, 1253, 1254, 1255]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|1|5|20.00%|[544, 545, 546, 547]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|3|14|21.43%|[114, 115, 256, 257, 259, 261, 262, 305, 306, 338, 349]|
|:large_blue_circle:|com/starrocks/connector/paimon/RowDataConverter.java|10|12|83.33%|[57, 63]|
|:large_blue_circle:|com/starrocks/catalog/PaimonTable.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java [hunk 1]
```diff
@@ -150,11 +152,6 @@ private static void processBaseTables(QueryStatement queryStatement, Set<BaseTab
                 continue;
             }
 
-            if (!FeConstants.isReplayFromQueryDump && isExternalTableFromResource(table)) {
```
Comment (QingdongZeng3) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:153]: Why delete this?

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 43 / 82 (52.44%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|1|18|05.56%|[1303, 1304, 1305, 1306, 1309, 1312, 1313, 1316, 1317, 1318, 1320, 1321, 1325, 1326, 1327, 1328, 1329]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|1|5|20.00%|[544, 545, 546, 547]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|3|14|21.43%|[116, 117, 258, 259, 261, 263, 264, 307, 308, 340, 351]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|20|25|80.00%|[716, 834, 885, 886, 887]|
|:large_blue_circle:|com/starrocks/connector/paimon/RowDataConverter.java|10|12|83.33%|[57, 63]|
|:large_blue_circle:|com/starrocks/catalog/PaimonTable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/PaimonPartitionKey.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 75 / 82 (91.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/paimon/RowDataConverter.java|10|12|83.33%|[57, 63]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|12|14|85.71%|[261, 351]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|16|18|88.89%|[1306, 1313]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|24|25|96.00%|[834]|
|:large_blue_circle:|com/starrocks/catalog/PaimonTable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/PaimonPartitionKey.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 79 / 86 (91.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/paimon/RowDataConverter.java|10|12|83.33%|[57, 63]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|12|14|85.71%|[261, 351]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|16|18|88.89%|[1306, 1313]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|28|29|96.55%|[843]|
|:large_blue_circle:|com/starrocks/catalog/PaimonTable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/PaimonPartitionKey.java|3|3|100.00%|[]|

**Review by miomiocat — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java [hunk 2]
```diff
@@ -177,6 +180,10 @@ private static boolean isExternalTableFromResource(Table table) {
             IcebergTable icebergTable = (IcebergTable) table;
             String catalogName = icebergTable.getCatalogName();
             return Strings.isBlank(catalogName) || isResourceMappingCatalog(catalogName);
+        } else if (table instanceof PaimonTable) {
+            PaimonTable icebergTable = (PaimonTable) table;
+            String catalogName = icebergTable.getCatalogName();
```
Comment (miomiocat) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:185]: paimonTable

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java [hunk 3]
```diff
@@ -818,6 +827,27 @@ private void checkPartitionColumnWithBaseIcebergTable(SlotRef slotRef, IcebergTa
             }
         }
 
+        @VisibleForTesting
+        void checkPartitionColumnWithBasePaimonTable(SlotRef slotRef, PaimonTable table) {
```
Comment (miomiocat) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:831]: private

File: fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java [hunk 4]
```diff
@@ -331,7 +348,7 @@ public static Map<String, List<List<String>>> getPartitionList(Table table, Colu
             throws UserException {
         if (table.isNativeTableOrMaterializedView()) {
             return ((OlapTable) table).getListPartitionMap();
-        } else if (table.isHiveTable() || table.isHudiTable() || table.isIcebergTable()) {
+        } else if (table.isHiveTable() || table.isHudiTable() || table.isIcebergTable() || table.isPaimonTable()) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java:351]: There are too many conditions to check, maybe we can refactor this to be better extensible later.

**Review by miomiocat — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 79 / 86 (91.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/paimon/RowDataConverter.java|10|12|83.33%|[57, 63]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|12|14|85.71%|[261, 351]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|16|18|88.89%|[1316, 1323]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|28|29|96.55%|[843]|
|:large_blue_circle:|com/starrocks/catalog/PaimonTable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/PaimonPartitionKey.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by Youngwb — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java (hunk 4 above)
Comment (wg1026688210) [reply fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java:351]: Next pr will work on it

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-3.2

**Pull request merged**